crates/ruma-events/src/room/message/content_serde.rs

```diff
@@ -76,6 +77,64 @@ impl<'de> Deserialize<'de> for MessageType {
     }
 }
 
+/// Helper struct to determine the itemtype from a `serde_json::value::RawValue`
+#[derive(Debug, Deserialize)]
+struct ItemTypeDeHelper {
+    /// The item type field
+    itemtype: String,
+}
+
+impl<'de> Deserialize<'de> for GalleryItemType {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let json = Box::<RawJsonValue>::deserialize(deserializer)?;
+        let ItemTypeDeHelper { itemtype } = from_raw_json_value(&json)?;
+
+        Ok(match itemtype.as_ref() {
+            "m.audio" => Self::Audio(from_raw_json_value(&json)?),
+            "m.file" => Self::File(from_raw_json_value(&json)?),
+            "m.image" => Self::Image(from_raw_json_value(&json)?),
+            "m.video" => Self::Video(from_raw_json_value(&json)?),
+            _ => Self::_Custom(from_raw_json_value(&json)?),
+        })
+    }
+}
+
+impl Serialize for GalleryItemType {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut map = match self {
+            GalleryItemType::Audio(content) => {
+                serde_json::to_value(content).map_err(serde::ser::Error::custom)?
+            }
+            GalleryItemType::File(content) => {
+                serde_json::to_value(content).map_err(serde::ser::Error::custom)?
+            }
+            GalleryItemType::Image(content) => {
+                serde_json::to_value(content).map_err(serde::ser::Error::custom)?
+            }
+            GalleryItemType::Video(content) => {
+                serde_json::to_value(content).map_err(serde::ser::Error::custom)?
+            }
+            GalleryItemType::_Custom(content) => {
+                serde_json::to_value(content).map_err(serde::ser::Error::custom)?
+            }
+        }
+        .as_object()
+        .cloned()
+        .unwrap_or_default();
+
+        map.insert("itemtype".to_string(), Value::String(self.itemtype().to_string()));
+        map.remove("msgtype");
+
+        map.serialize(serializer)
```

Comment:
This sort of sucks. I wanted to avoid having to duplicate all the content types but still be able to serialize `msgtype` as `itemtype`. Curious if anyone has better ideas on how to achieve this.